2020-05-16-password-2.0.md

```diff
@@ -1,10 +1,246 @@
+    (from [aeson](https://hackage.haskell.org/package/aeson)),
+    [FromHttpApiData](https://hackage.haskell.org/package/password-instances-2.0.0.1/docs/Data-Password-Instances.html#v:-36-fFromHttpApiDataPassword)
+    (from [http-api-data](https://hackage.haskell.org/package/http-api-data)),
+    and
+    [PersistField](https://hackage.haskell.org/package/password-instances-2.0.0.1/docs/Data-Password-Instances.html#v:-36-fPersistFieldPasswordHash)
+    (from [persistent](https://hackage.haskell.org/package/persistent)).  These
+    instances are frequently used when doing actual web development.
+
+-   Great documentation.  Anything confusing in the Haddocks is considered a
+    "high-priority" bug.
+
+## What changed in `password-2.0`?
+
+`password-2.0` is a complete rewrite of the `password` library by
+[Felix Paulusma](https://github.com/Vlix)[^1].  The architecture of the library
+has been completely redone to support multiple hashing algorithms.
+
+The following algorithms are now supported:
+
+-   [Argon2](https://hackage.haskell.org/package/password-2.0.1.1/docs/Data-Password-Argon2.html)
+-   [Bcrypt](https://hackage.haskell.org/package/password-2.0.1.1/docs/Data-Password-Bcrypt.html)
+-   [PBKDF2](https://hackage.haskell.org/package/password-2.0.1.1/docs/Data-Password-PBKDF2.html)
+-   [Scrypt](https://hackage.haskell.org/package/password-2.0.1.1/docs/Data-Password-Scrypt.html)
+    (this was the only algorithm supported in the original `password-1.0` library)
+
+`password-2.0` also adds a bunch of much-needed documentation, as well as many
+tests.
+
+## How to use `password`?
+
+This section will present a short tutorial for using `password`.  The following
+commands are run in a GHCi session.  You can follow along if you start GHCi
+with a command like `stack --resolver nightly-2020-05-17 ghci --package
+password-2.0.1.1`.
+
+We enable a couple extensions to make the following code a little easier:
+
+```haskell
+> :set -XOverloadedStrings
+```
+
+The first step in using `password` is to pick which hashing algorithm you want to use.
+If you don't know which to pick, we recommend
+[Bcrypt](https://hackage.haskell.org/package/password-2.0.1.1/docs/Data-Password-Bcrypt.html).
+The trade-offs of the different algorithms are discussed in more detail
+[in the Haddocks](https://hackage.haskell.org/package/password-2.0.1.1/docs/Data-Password.html).
+
+Once you've picked the hashing algorithm, you can import its module.  We use
+[`Data.Password.Bcrypt`](https://hackage.haskell.org/package/password-2.0.1.1/docs/Data-Password-Bcrypt.html)
+as an example:
+
+```haskell
+> import Data.Password.Bcrypt
+```
+
+Next, we need a
+[`Password`](https://hackage.haskell.org/package/password-2.0.1.1/docs/Data-Password-Bcrypt.html#t:Password)
+to work with.  Let's create one with
+[`mkPassword`](https://hackage.haskell.org/package/password-2.0.1.1/docs/Data-Password-Bcrypt.html#v:mkPassword)[^2]:
+
+```haskell
+> let pass = mkPassword "foobar"
```

Comment:
Seeing as `OverloadedStrings` is on, you could also mention the `IsString` instance of `Password`, but it's not very important. I'm also not sure if it even necessary to mention in this post.

Reply:
Yeah, I imagine that the `IsString` instance probably wouldn't be used too much (except for maybe testing), so I didn't think it was too important to bring it up here.